### DISCUSSION TITLE
Some processes don't register in list-apps or list-windows

### DISCUSSION THREAD

**Discussion by globalshrug** [nikitabobko/AeroSpace#991] [potential-bugs]
Some processes/windows don't get picked up by Aerospace.

This is not _that_ surprising for some processes - I encountered this with `mpv` and given that it doesn't even come close to using a macOS-native GUI, it's not shocking. But, `IINA` is supposed to be a macOS-native GUI frontend for `mpv` and also doesn't show up.

With both apps showing windows in the background, neither show up in aerospace's query commands:

```
$ iina && mpv --player-operation-mode=pseudo-gui & aerospace list-apps && echo && aerospace list-windows --all
[1] 13242
14266 | com.apple.finder                  | Finder
14437 | com.apple.Safari                  | Safari
14720 | net.kovidgoyal.kitty              | kitty
75355 | com.apple.iCal                    | Calendar
75783 | com.apple.Music                   | Music
81677 | com.apple.mail                    | Mail
85755 | com.microsoft.rdc.macos           | Microsoft Remote Desktop
59055 | com.apple.Notes                   | Notes
25845 | com.apple.MobileSMS               | Messages
34649 | com.google.Chrome                 | Google Chrome
38343 | com.apple.Preview                 | Preview

5297  | Calendar | Calendar
14962 | Finder   | Desktop
14911 | Mail     | Yesterday – 2 messages
1371  | Safari   | New Issue · nikitabobko/AeroSpace
1552  | kitty    | tmux - zsh
$
```

## Checklist

- [ X ] I've searched for duplicates. My search queries were (please don't forget to search in closed issues too):
  - list-apps
  - list-windows
  - ...
- [ X ] I've read the guide https://nikitabobko.github.io/AeroSpace/guide
- [ X ] I've searched in the documentation (https://nikitabobko.github.io/AeroSpace/guide & https://nikitabobko.github.io/AeroSpace/commands). My search queries were:
  - process
  - window
  - ...

```
aerospace CLI client version: 0.14.2-Beta 0cb8dbdfc5ee73b8cbc200f175f467ebead55201
AeroSpace.app server version: 0.14.2-Beta 0cb8dbdfc5ee73b8cbc200f175f467ebead55201
```

**Comment by globalshrug** [comment]
Looks like there's a difference between launching a process via the command-line versus either Finder, or directing Finder to open the app from the command line. E.g. `$ open -a vlc` vs. `$ vlc` will produce different results from Aerospace's perspective. The former produces an app and window being registered by Aerospace, but the latter doesn't. This works for VLC and IINA, but not mpv since `open -a mpv` doesn't run the process.

**Comment by quulah** [comment]
I just noticed the same. I was wondering why few of my windows didn't get tiled properly. This has happened a few times before, but now I by chance checked `list-windows`. Neither was there.

Restarting seems to have the issue once more. I tend to always start these two from the Dock, so there's probably something else at play as well than just the method of starting.

The two in question, WezTerm and Neovide, do seem like well-behaved macOS apps as well. :)

Now that I thought of this, I experimented a bit. A theory is that switching between an external monitor and the laptop's has something to do with this. Again, some windows failed to be tiled automatically, but they did show up in the `list-windows` output this time so I'm not sure anymore whether I'm experiencing the same issue or not. Or perhaps there's a few different ones with the same result.

**Comment by globalshrug** [comment]
@quulah I've started working around this a bit by pretending `mpv` is a cocoa app.

I followed the instructions here: [https://tmewett.com/making-macos-bundle-info-plist/](https://tmewett.com/making-macos-bundle-info-plist/), and hard-linked the `mpv` binary in. I haven't fully got the iconset working yet, but everything else seems to work. Add the following into your `~/.config/mpv/mpv.conf` to make `mpv` pretend to be a normal app (see: [https://mpv.io/manual/master/#pseudo-gui-mode](https://mpv.io/manual/master/#pseudo-gui-mode)):

`force-window=yes`
`idle=once`
`no-terminal`

Once I'm sure the bundle is working properly, I might push it up to a repo.

I am still seeing some weird behaviour with window placement and scaling, but that might come down to Aerospace fighting to keep the window layout versus mpv trying to adjust its window to the right size for the video, or something. I found that it works best just to float the window.

Update: calling it from the command line works best like so: `open -a mpv.app`, since `mpv` will still be linked to whichever `mpv` binary comes first in your $PATH. You want the bundle, not the binary, and .app makes `open` look for that one instead.

**Comment by bassamsdata** [comment]
I hope you don't mind me commenting on this issue since I’m experiencing the same issue with Neovide when launching it from the terminal. I can’t even get the app to respond to any Aerospace shortcuts, almost like it doesn’t recognize the app at all.

the `aerospace list-windows --all` command sees neovide.
```sh
945 | Arc     | Some processes don't register.
147 | WezTerm |  ~/.config/nvim
113 | WezTerm |  ~/repos/projectsPort/newFormatCL
956 | neovide | Neovide
966 | neovide | Neovide
```
but `aerospace list-apps` doesn't as it sees neovide as NULL

```sh
444  | com.apple.finder           | Finder
831  | com.github.wez.wezterm     | WezTerm
9365 | company.thebrowser.Browser | Arc
9527 | NULL-APP-BUNDLE-ID         | neovide
9711 | NULL-APP-BUNDLE-ID         | neovide
```

**Comment by globalshrug** [comment]
> I hope you don't mind me commenting on this issue since I’m experiencing the same issue with Neovide when launching it from the terminal. I can’t even get the app to respond to any Aerospace shortcuts, almost like it doesn’t recognize the app at all.
> 
> the `aerospace list-windows --all` command sees neovide.
> 
> ```shell
> 945 | Arc     | Some processes don't register.
> 147 | WezTerm |  ~/.config/nvim
> 113 | WezTerm |  ~/repos/projectsPort/newFormatCL
> 956 | neovide | Neovide
> 966 | neovide | Neovide
> ```
> 
> but `aerospace list-apps` doesn't as it sees neovide as NULL
> 
> ```shell
> 444  | com.apple.finder           | Finder
> 831  | com.github.wez.wezterm     | WezTerm
> 9365 | company.thebrowser.Browser | Arc
> 9527 | NULL-APP-BUNDLE-ID         | neovide
> 9711 | NULL-APP-BUNDLE-ID         | neovide
> ```

Yeah, Aerospace needs a bundle ID, I think. If you follow the steps I posted, you might have some luck. Otherwise I can try to give it a Bundle ID tomorrow like I did with mpv and see if it behaves differently.

**Comment by bassamsdata** [comment]
@ashleyharvey  Thanks for the quick response and helpful info! It was a bit of a hectic process since this was my first time doing something like this. 

In the end, it worked with just a few steps, thank you so much! It finally solved a problem I’ve been dealing with for a long time.

**Comment by johnend** [comment]
This seems to be occurring with me on the most recent update. Previously Ghostty, Arc, Spark and Outlook would be adopted by aerospace, but not they aren't. None of those three while open register in `list-windows --all` or `list-apps`.

![image](https://github.com/user-attachments/assets/f0485b6a-2862-45bc-a165-3e92f5149194)

This is running in Ghostty and Ghostty isn't listed. Though Arc, Spark and Outlook also should be listed. Previously all three of these worked.

**Comment by kzssc** [comment]
I'm unsure if it's related, but lately I have been experiencing a rather violent flickering that happens only to Spark (2.11.56, com.readdle.smartemail-Mac) upon wake. No other app does this, and it doesn't matter whether Spark was in focus or not before sleep. If spark is not in focus, then it'll hide after the flickers. very weird. no resizing is done. 

As a way to circumvent this I tried "layout floating" and ""macos-native-fullscreen on", though neither helped. 

 I'm not 100% sure of the cause of this, though fairly confident that it didn't happen in previous versions. 

any insight? 

perhaps this is relevant. https://github.com/nikitabobko/AeroSpace/issues/1216 behaviour very similar to this: https://github.com/nikitabobko/AeroSpace/issues/924
